espnet2/enh/separator/asteroid_models.py

```diff
@@ -0,0 +1,144 @@
+
+from espnet2.enh.separator.abs_separator import AbsSeparator
+
+
+class AsteroidModel_Converter(AbsSeparator):
+    def __init__(
+        self,
+        encoder_output_dim: int,
+        model_name: str,
+        num_spk: int,
+        pretrained_path: str = "",
+        loss_type: str = "si_snr",
+        **model_related_kwargs,
+    ):
+        """The class to convert the models from asteroid to AbsSeprator.
+
+        Args:
+            encoder_output_dim: input feature dimension, deafult=1 after the NullEncoder
+            num_spk: number of speakers
+            loss_type: loss type of enhancement
+            model_name: asteroind model names, e.g. ConvTasNet, DPTNet. Refers to
+                        https://github.com/asteroid-team/asteroid/
+                        blob/master/asteroid/models/__init__.py
+            pretrained_path: the name of pretrained model from Asteroid in Zenodo.
+                             Refers to: https://github.com/asteroid-team/asteroid/
+                             blob/master/docs/source/readmes/pretrained_models.md
+            model_related_kwargs: more args towards each specific asteroid model.
+        """
+        super(AsteroidModel_Converter, self).__init__()
+
+        assert (
+            encoder_output_dim == 1
+        ), encoder_output_dim  # The input should in raw-wave domain.
+
+        # Please make sure the installation of Asteroid.
+        # https://github.com/asteroid-team/asteroid
+        from asteroid import models
+
+        model_related_kwargs = {
+            k: None if v == "None" else v for k, v in model_related_kwargs.items()
+        }
+        # print('args:',model_related_kwargs)
+
+        if pretrained_path:
+            model = eval(
+                "models.{}.from_pretrained('{}')".format(model_name, pretrained_path)
+            )
+        else:
+            model_name = getattr(models, model_name)
+            model = model_name(**model_related_kwargs)
+
+        self.model = model
+        self._num_spk = num_spk
+
+        self.loss_type = loss_type
+        if loss_type != "si_snr":
+            raise ValueError("Unsupported loss type: %s" % loss_type)
+
+    def forward(self, input: torch.Tensor, ilens: torch.Tensor = None):
+        """Whole forward of asteroid models.
+
+        Args:
+            input (torch.Tensor): Raw Waveforms [B, T]
+            ilens (torch.Tensor): input lengths [B]
+
+        Returns:
+            estimated Waveforms(List[Union(torch.Tensor]): [(B, T), ...]
+            ilens (torch.Tensor): (B,)
+            others predicted data, e.g. masks: OrderedDict[
+                'mask_spk1': torch.Tensor(Batch, T),
+                'mask_spk2': torch.Tensor(Batch, T),
+                ...
+                'mask_spkn': torch.Tensor(Batch, T),
+            ]
+        """
+
+        # import soundfile as sf
+        # sf.write('test_mix.wav',input.data.cpu()[0],8000)
+        est_source = self.model(input)  # B,nspk,T or nspk,T
```

Comment:
This will work for waveform to waveform models, but we actually have `forward_wav` for models which are not trained this way e.g. deep clustering that we should add soon.